### DISCUSSION TITLE
tabula-py and Python 3.12

### DISCUSSION THREAD

**Discussion by jflaatsch** [chezou/tabula-py#365] [Q&A] [answered]
Is tabula-py compatible with Python 3.12?  I haven't been able to install it using pip on Windows 11.  I followed multiple leads across the web.  Please let me know if I'm missing something obvious or if I just need to wait for a minor update.

I have the following installed:
<code>
c:\>python --version
Python 3.12.0

c:\>pip --version
pip 23.3.1 from C:\Users\Jonathan\AppData\Local\Programs\Python\Python312\Lib\site-packages\pip (python 3.12)

c:\>java --version
java 21.0.1 2023-10-17 LTS
Java(TM) SE Runtime Environment (build 21.0.1+12-LTS-29)
Java HotSpot(TM) 64-Bit Server VM (build 21.0.1+12-LTS-29, mixed mode, sharing)

c:\>cmake --version
cmake version 3.27.7

CMake suite maintained and supported by Kitware (kitware.com/cmake).

c:\>echo %PATH%
C:\Program Files\Common Files\Oracle\Java\javapath;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Users\Jonathan\AppData\Local\Programs\Python\Python312;C:\Program Files\Microsoft SQL Server\120\Tools\Binn\;C:\Program Files (x86)\QuickTime\QTSystem\;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Program Files\Java\jdk-21\bin;C:\Users\Jonathan\AppData\Local\Programs\Python\Python312\Scripts\;C:\Users\Jonathan\AppData\Local\Programs\Python\Python312\;C:\Users\Jonathan\AppData\Local\Microsoft\WindowsApps;C:\Program Files (x86)\Windows Kits\10\bin\x64;
</code>

I have <b>Visual Studio Build Tools 2022 - 17.7.6</b> with "<b>Desktop development with C++</b>" and <b>Visual Studio Community 2022 - 17.6.7</b> with "Python development", "Node.js development", "<b>Desktop development with C++</b>", and "Linux and embedded development with C++
".  

The error while installing tabula-py is the following:
<code>
 Building wheel for jpype1 (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [127 lines of output]
...
      jp_primitivetype.cpp
      native\common\jp_primitivetype.cpp(50): error C2039: 'ob_digit': is not a member of '_longobject'
      C:\Users\Jonathan\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
      native\common\jp_primitivetype.cpp(50): error C2039: 'ob_digit': is not a member of '_longobject'
      C:\Users\Jonathan\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for jpype1
  Running setup.py clean for jpype1
Failed to build jpype1
ERROR: Could not build wheels for jpype1, which is required to install pyproject.toml-based projects
</code>

This may be related to similar issues installing packages with Python 3.12, or something else with compiling in C++: 
- https://github.com/KillianLucas/open-interpreter/issues/581
- https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2039?view=msvc-170

Any suggestions to work the issue or requests for additional information are welcome.

**Comment by chezou** [comment]
Thanks for reporting. It looks like jpype doesn't support 3.12 yet.
https://github.com/jpype-project/jpype/issues/1147
https://github.com/jpype-project/jpype/pull/1148

I believe once jpype supports 3.12, it will be fixed.

**Reply by jflaatsch** [reply]
Thanks for the quick response, Yes, it seems there are some issues between jpype1 and python 3.12.  
I'll wait for a fix since my limited experience with Java and Python will be no help at all. :)

**Comment by chezou** [comment, marked as answer]
I released v2.9.0rc0 to avoid jpype installation error on Python 3.12. https://pypi.org/project/tabula-py/2.9.0rc0/ Please try it and give me any feedback 😄
